8 new actions rework

kaumudpa opened the pull request:
Rework to replace PR #51 

@redDwarf03 This should resolve the issues we were facing.


--PR DETAILS--

Did some corrections in `utils_js.dart` to make sure `readContract`, r`eadContracts`  & `getChains` work. (No idea why this was broken)

Also implemented

`getChains`
`disconnect`
`switchAccount`
`switchChain`
`verifyMessage`
`waitForTransactionReceipt`
`watchAccount`
`watchConnections`

@Chralu / @redDwarf03  Requesting you to kindly review these and merge this if everything looks ok.

redDwarf03 commented:
Tests
- `getChains`: ok
- `disconnect: ko
When i click on Disconnect Wallet in example app (after Connect wallet), i meet this error:
```
[  +12 ms] TypeError: connector.disconnect is not a function
           assets/packages/wagmi_flutter_web/assets/main.js 51204:25                         disconnect
           assets/packages/wagmi_flutter_web/assets/main.js 25834:92                         <fn>
           assets/packages/wagmi_flutter_web/assets/main.js 25845:30                         disconnect
           packages/wagmi_flutter_web/src/wagmi_core.dart 355:11                             disconnect
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
           packages/wagmi_flutter_web/src/wagmi_core.dart 351:33                             disconnect
           packages/example/main.dart 124:34                                                 <fn>
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
           packages/example/main.dart 119:26                                                 <fn>
           packages/flutter/src/material/ink_well.dart 1171:21                               handleTap
           packages/flutter/src/gestures/recognizer.dart 344:24                              invokeCallback
           packages/flutter/src/gestures/tap.dart 652:11                                     handleTapUp
           packages/flutter/src/gestures/tap.dart 309:5                                      [_checkUp]
           packages/flutter/src/gestures/tap.dart 242:7                                      handlePrimaryPointer
           packages/flutter/src/gestures/recognizer.dart 696:9                               handleEvent
           packages/flutter/src/gestures/pointer_router.dart 98:7                            [_dispatch]
           packages/flutter/src/gestures/pointer_router.dart 143:9                           <fn>
           dart-sdk/lib/_internal/js_dev_runtime/private/linked_hash_map.dart 21:7           forEach
           packages/flutter/src/gestures/pointer_router.dart 141:17                          [_dispatchEventToRoutes]
           packages/flutter/src/gestures/pointer_router.dart 127:7                           route
           packages/flutter/src/gestures/binding.dart 495:5                                  handleEvent
           packages/flutter/src/gestures/binding.dart 475:14                                 dispatchEvent
           packages/flutter/src/rendering/binding.dart 425:11                                dispatchEvent
           packages/flutter/src/gestures/binding.dart 420:7                                  [_handlePointerEventImmediately]
```

kaumudpa commented:
You are connected via Wallet Connect protocol or injected metamask? 

When we connect in a browser with metamask installed - the connection type is `injected` but if we are connected via QR code or deeplink the type is `walletConnect` 

Please tell me how are you connected?

redDwarf03 commented:
> @redDwarf03 You are connected via Wallet Connect protocol or injected metamask?
> 
> When we connect in a browser with metamask installed - the connection type is `injected` but if we are connected via QR code or deeplink the type is `walletConnect`
> 
> Please tell me how are you connected?

cf my new message. i met issues with Frame. But with Rabby, it's ok

kaumudpa commented:
Yes got the issue - resolving it. Please test the rest of the actions in the mean time, Will push a commit shortly.

redDwarf03 commented:
> * Switch Account / Switch chain : a popup with info appears but i don't understand the goal of the example app for these methods (could you explain how you test these method please)
> * Watch Account, Watch connections : i click on watch account and I change my account in rabby for example, and nothing is displayed on the example app. i don't know how to test these method too.. sorry

like i said.. i don't understand the use cases to test this:

- Switch Account / Switch chain : a popup with info appears but i don't understand the goal of the example app for these methods (could you explain how you test these method please)
- Watch Account, Watch connections : i click on watch account and I change my account in rabby for example, and nothing is displayed on the example app. i don't know how to test these method too.. sorry

If you have any helps...

redDwarf03 commented:
wait for tx receipt. What if the chain id for hash 0x041f2da24620eeef645e646f07399b3b374b86201931b228578c4632fb41abf4 ?

kaumudpa commented:
Allow me a few hrs - let me write proper use cases here for you to test

kaumudpa commented:
> 0x041f2da24620eeef645e646f07399b3b374b86201931b228578c4632fb41abf4

Amoy

redDwarf03 commented:
> > 0x041f2da24620eeef645e646f07399b3b374b86201931b228578c4632fb41abf4
> 
> Amoy - 80002

wagmi_core.ts:52 Fetch API cannot load wss://polygon-amoy.g.alchemy.com/v2/DynWKvz6PUFaeZNmlxPXNiV1nK4Ac_2D. URL scheme "wss" is not supported.

redDwarf03 commented on example/lib/context.dart:
```diff
@@ -89,7 +89,7 @@ class WagmiContext {
       return 'wss://polygon-amoy.g.alchemy.com/v2/DynWKvz6PUFaeZNmlxPXNiV1nK4Ac_2D';
```
wss is not compatible with transport.http

kaumudpa replied:
Will resolve it as well in the upcoming commit

kaumudpa replied:
I will also post a solution here tomorrow where we have implemented using both wss and http together with fallback mechanism as mentioned here - https://wagmi.sh/core/api/transports/fallback

kaumudpa commented:
Hi @redDwarf03 - The correction you requested are now implemented.

1. `disconnect` is working - Connected walled gets disconnected.
2. `getChains ` - lists the chains available.
3. `switchAccount ` - This helps in switching accounts if multiple accounts are connected via multi config. (You wont be able to test it rn as multiconfig still needs some work, but its working)
4. `switchChain ` - shows a dialogue to change the chain of the connected wallet - To check this connect the wallet, click on  `get account info` and then click on `switch chain`.
5. `verifyMessage ` - Verifies the signed message.
6. `waitForTransactionReceipt ` - provided a tx it waits for it to be included in the block and returns the tx receipt.
7. `watchAccount ` - When we click on watch account it subscribes to account changes and whenever a wallet connect/disconnects or changes it notifies. To test click on `watch account` button and try connecting/discoonecting a wallet, whenever you do that changes are emitted on the console.
8. `watchConnections ` similar to `watchAccount` but instead of subscribing to just account changes it subscribes to the whole connection object. Test it similarly like `watchAccount`.

feat.

Also implemented
9. `watchAsset` - Try using it - When provided with a token contract address, decimal and symbol - it asks the connected wallet (metamask) to add that token to the account to watch.

Please check and let me know.



-----Minor Bug-------
`disconnect` - This is working fine when the connected wallet is `injected` - there is some error with `walletConnect` type connection, maybe @Chralu can help with it.

-----Issues since multi config-----
You notice the changes i did in `wagmi_flutter_web.dart` - thats to resolve the issue i mentioned in #55, this issue is there since we introduced multiconfig Even my solution is not the right solution, it does solves the ability to hot reload but sometimes the config is `null` and even if the reload is successful the connected wallet's state (config) is lost due to reload. I was hoping if @Chralu  can look into it.

redDwarf03 commented:
Hi @kaumudpa 
i have some trouble to test all features. When i connect my Rabby Wallet, wallet connect modal is ok but get account info shows account address unknown, status disconnected, etc... (see image). 
![Capture d’écran 2024-09-23 à 09 23 03](https://github.com/user-attachments/assets/a757eff7-66c3-4ab0-8f27-7576a4f235b1)


how come the wallet is connected but the object account is not filled in, in your opinion? This causes errors in the functions that require the account object.

kaumudpa commented:
@redDwarf03 Its because of the #55 issue, Try reverting `wagmi_flutter_web.dart` to the exact code available in main branch.

kaumudpa commented:
@redDwarf03 This issue is since we introduced multi-config, we need to resolve this - with my implementation of `wagmi_flutter_web.dart` We are able to hot reload/reload but we face the problem you are getting and if we use the main branchs code for `wagmi_flutter_web.dart` hotreload/reload stops - #55

redDwarf03 commented:
Same issue with `wagmi_flutter_web` updated. i will discuss with @Chralu about this

kaumudpa commented:
Can we connect on Discord / Slack / Telegram to collaborate efficiently?

Chralu commented:
Hi,

I'll check that reload issue. 

In the meanwhile, as a workaround I suggest using hot-restart and refreshing the page in the browser. 
When I was working on the lib, it was the way I used.

kaumudpa commented:
> Hi,
> 
> I'll check that reload issue.
> 
> In the meanwhile, as a workaround I suggest using hot-restart and refreshing the page in the browser. When I was working on the lib, it was the way I used.

Even hot-restart gets stuck on loader, and if e manually refresh the page in browser we face the problem @redDwarf03 is facing.

kaumudpa commented:
@Chralu 

Some important bugs to consider while you look into this.

Since multiconfig merge the following things are broken.

1. If a wallet is connected and the app is reloaded or manually refreshed in the browser, the connection state is lost and sometimes the app asks again to connect the wallet, While before multiconfig the conneced wallet state was retained.

2. Since multiconfig - even if wallet is connected successfully via walletConnect it always returns a disconnected state. - So if we connect a wallet via walletconnect QR and get account info it shows the state to be disconnected.  Before multiconfig this was working fine.

kaumudpa commented:
I guess creating initialising config object in dart and then passing it to JS needs a relook. Before multiconfig we were doing this directly at JS side.

kaumudpa commented:
@Chralu Any updates from your end regarding this?

kaumudpa commented:
@Chralu  we have been working on this since last 5 days round the clock, finally found a solution, will post it here tomorrow for you to review.

kaumudpa commented:
Hi @Chralu & @redDwarf03 It took us multiple days of experiment and strong testing to finally all the issues we were facing here and in #55. The root cause was as i mentioned earlier in the multi config approach. To resolve this we will now have to create new configs with `wagmi.Web3Modal.createConfig` and with a parameter `configKey` which will be a unique id string for every config and then using this key in every method we want to call by passing this key as another param in that method.

Rest default config remains the sam, If no key or wrong key is provided to a method it will use default config.

We have tested this extensively and everything is finally working fine now, This PR will not break anything.

I request @Chralu to please review this and merge this PR at the earliest.

kaumudpa commented:
@Chralu & @redDwarf03 can you please check this, merge and create a new release on pub as the current release on pub is broken.

redDwarf03 commented:
Some features don't work when i try them, but this may depend on the way in which the test application is used, which is not yet fully complete. 
I'm going to publish your work on pub.dev so that you can move forward and not get stuck. We'll come up with solutions if we run into problems.
Sorry for the lack of availability these days, but there's a lot of work going on in parallel with this open source project.

kaumudpa commented:
@redDwarf03 We did tested everything multiple times, Just let me know what all problems you faced whenever you get time to go through this. Will be willing to correct it

redDwarf03 commented:
https://pub.dev/packages/wagmi_flutter_web available

- `switchAccount` ok
- `switchChain`ok
- 

Concerning my issues
- `disconnect`: i click on connect button, i choose Rabby, I click on Get Account info (the popup is ok) and i click on disconnect and i meet an error
```
[ +167 ms] TypeError: Cannot read properties of undefined (reading 'getProvider')
           assets/packages/wagmi_flutter_web/assets/main.js 2:1149613                        disconnect
           dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 369:37  _checkAndCall
           dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 431:39  dcall
           assets/packages/wagmi_flutter_web/assets/main.js 2:1135069                        n
           assets/packages/wagmi_flutter_web/assets/main.js 2:720839                         <fn>
           assets/packages/wagmi_flutter_web/assets/main.js 2:722136                         disconnect
           packages/wagmi_flutter_web/src/wagmi_core.dart 358:11                             disconnect
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
           packages/wagmi_flutter_web/src/wagmi_core.dart 354:33                             disconnect
           packages/example/main.dart 170:34                                                 <fn>
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
           dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
           packages/example/main.dart 165:26                                                 <fn>
           packages/flutter/src/material/ink_well.dart 1171:21                               handleTap
           packages/flutter/src/gestures/recognizer.dart 344:24                              invokeCallback
           packages/flutter/src/gestures/tap.dart 652:11                                     handleTapUp
           packages/flutter/src/gestures/tap.dart 309:5                                      [_checkUp]
           packages/flutter/src/gestures/tap.dart 242:7                                      handlePrimaryPointer
           packages/flutter/src/gestures/recognizer.dart 696:9                               handleEvent
           packages/flutter/src/gestures/pointer_router.dart 98:7                            [_dispatch]
           packages/flutter/src/gestures/pointer_router.dart 143:9                           <fn>
```
- `switchAccount`. When i switch on my wallet, an error occurs
```
[+88246 ms] DartError: NoSuchMethodError: ''
            Dynamic call with too many positional arguments. Expected: 1 Actual: 2
            Receiver: Instance of '(JSObject) => Null'
            Arguments: [Instance of 'LegacyJavaScriptObject', Instance of 'LegacyJavaScriptObject']
            dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 296:3       throw_
            dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 893:3   defaultNoSuchMethod
            dart-sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart 63:17                 noSuchMethod
            dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 868:31  noSuchMethod
            dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 318:12  callNSM
            dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 428:10  _checkAndCall
            dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 431:39  dcall
            assets/packages/wagmi_flutter_web/assets/main.js 2:1154725                        o
            assets/packages/wagmi_flutter_web/assets/main.js 2:1158644                        <fn>
```
- `watchAsset` i fill contractAddress of a token, the Symbol and the Decimal. when i click on add token, Rabby opens but freezed

redDwarf03 commented:
@kaumudpa Chralu will work on lib next tuesday normally

redDwarf03 commented:
Rabby with browser extension

kaumudpa commented:
@redDwarf03 Yes got it - The disconnect issue is since the beginning even if we try to disconnect it from inbuilt jsmodal here is how it occurs.

1. We connect the wallet.
2. We refresh/reload the page.
3. We connect the wallet again after refresh.
4. Now if we try to disconnect it cannot disconnect (errors).

On a new flutter run instance if we Connect the wallet for the very first time & disconnect it does disconnect perfectly. 

Or if disconnect the wallet before reload/refresh and then reconnect after reload/refresh and then disconnect it works perfectly.

The issue is when we try to reload/refresh the app when the wallet is already in a connected state and then reconnect the wallet after the reload/refresh. 

I will make a seperate issue to track it there.
